README
NAME
    SquidGuard Manager - Web GUI for SquidGuard and SquidClamav
    administration

DESCRIPTION
    SquidGuard Manager is a Perl CGI Web GUI for squidGuard and SquidClamav
    administration. It allows full management of SquidGuard blocklist and
    ACLs. You can also manage graphicaly your SquidClamav configuration.

    This program supports all configuration directives of SquidGuard and
    SquidClamav without intrusive modification, i-e you can still edit the
    configuration files by hand before and after using SquidGuard Manager.

FEATURES
  SquidGuard Manager GUI interfaces
    Globals
        This interface allows you to define SquidGuard global variables.

    Schedules
        This interface allows you to define schedules time in order to use
        in other SquidGuard objects (Sources, URL Rewrites, Filters and
        ACLs).

    Sources
        Use this interface to define sources like users, groups, Ip
        addresses, range of Ip addresses or networks to apply policies on.
        These can be issued from a list, a file, an ldap query, a mysql
        query or a command line utility.

    Url rewriting
        This interface allows you to define URL rewriting groups, each
        consists of one or more Sed style substitution rules. They can be
        applied to source groups in ACLs menu.

    Filters
        Destination Groups filter selector in other words. In this interface
        you can shape each group by selecting a list for domain, url or
        regexp filter from those you defined in Manage Lists interface.

    ACLs
        This is the central place where you define what source objects
        follow what filters and when. Here everything fits together.

    View Configuration
        This interface allows you to see the SquidGuard configuration file.

    Manage Lists
        This interface allows you to build blocklists for each destination
        group. You can edit a whole list or add/remove individual entries.

    View logs
        This menu allows you to consult all log files in the log directory
        you'd defined in the SquidGuard global configuration.

    SquidClamav
        This link leads you to the GUI interface to administer SquidClamav
        configuration file. This interface is enabled only if you have
        SquidClamav installed on your system.

    Restart Squid
        This link allows you to 'reconfigure' or quick-restart Squid to make
        it recognize the changes you made to SquidGuard and/or SquidClamav
        configuration.

  SquidClamav Manager GUI interfaces
    Since release 1.4 SquidClamav Manager is able to manage both SquidClamav
    v5.x and v6.x branch.

    Globals
        This interface allows you to manage SquidClamav global variables.

    Virus Scanning
        This interface allows you to exclude URL or Content-Type from virus
        scan by defining regexp patterns.

    Whitelists
        Use this interface to define which domain name or site hostname to
        be whitelisted, i-e there happens no virus scan nor call to
        SquidGuard on these sites.

    Trusted Users
        Use this interface to define which users (RFC-931 usernames) will be
        excluded from virus scan and SquidGuard blocklists.

    Trusted Clients
        Use this interface to define which client hostnames or ip addresses
        will be excluded from virus scan and SquidGuard blocklists.

    View Configuration
        This interface allows you to see the SquidClamav configuration file.

    View log
        This interface allows you to consult the SquidClamav log file.

    In resume SquidGuard Manager is capable of every configuration task on
    SquidGuard and SquidClamav which otherwise you have to perform by
    opening terminal, editing files, becoming root and running commands
    manually. Modifications are written immediately to the configuration
    files.

REQUIREMENT
    SquidGuard Manager may run on any Unix like OS with any modern Perl
    installation using the perl CGI module. It must be run on a proxy server
    hosting squidGuard.

    You need to run this CGI script using Apache HTTP server. And the httpd
    daemon must be run as uid and gid holding SquidGuard (most of the time
    this is Squid cache effective user/group). This allows SquidGuard
    Manager to read and write the configuration files to manage blocklists
    and acls. If you use SquidClamav, it must be run under the same user
    too.

    SquidGuard Manager uses some system commands to manage blocklists and
    log files. Your system must have the following binaries installed: diff,
    grep, find, rm and tail.

INSTALLATION
    To install SquidGuard Manager you must be able to gain root privilege so
    that installer script can put files with appropriate permission.

  Prepare your system
    SquidGuard Manager needs to have read/write access onto SquidGuard
    configuration file and the dbhome directory. It must also have read
    rights onto the log file. If you use the SquidClamav Manager GUI it also
    requires read/write access to the squidclamav.conf file as well as the
    log file.

    For that reason SquidGuard Manager must be run as the user under which
    SquidGuard and SquidClamav are run. Most of the time this is the Squid
    cache effective user.

            chown -R squid:squid /usr/local/squidGuard
    or
            chown -R squid:squid /etc/squidguard.conf /var/lib/squidguard

    To allow httpd to run squidguardmgr.cgi as appropriate user/group you
    must edit your httpd.conf file and set the User and Group directives to
    the user and group running Squid.

    After configuration changes squid ought to be 'reconfigure'd, in other
    words quick-restarted. SquidGuard Manager Web GUI does it by using a
    'setuid root'ed wrapper. If your system doesn't support the feature
    unfortunately, you may need to do that by hand.

    IMPORTANT: SquidGuard Manager doesn't support symbolic links in the
    SquidGuard dbhome directory. If it is your case, you need to remove
    those symbolic links and modify your squidguard.conf file so that refers
    to the real files and directories.

  Automatic installation
    Execute the install script and answer to some questions it asks:

            $ perl Makefile.PL
            $ make
            $ su
            # make install

    By default it will install all files into /var/www/squidguardmgr/ and
    show you a configuration sample that should be included in your
    httpd.conf. Also, you can find some Apache configuration samples in
    contrib/ directory of the SquidGuard Manager source repository. Once you
    have done, restart httpd server and take a look at
    http://my.proxy.dom/squidguardmgr/ or
    http://my.proxy.dom:port/squidguardmgr/

    You can fully customize this installation by passing the following
    variables to Makefile.PL:

            SQUIDGUARD => Absolute path to the SquidGuard executable. Set to
                         'off' to disable its support
            CONFFILE => Absolute path to the SquidGuard configuration file
            DBHOME => Absolute path to the SquidGuard DB directory
            LOGDIR => Absolute path to the SquidGuard log directory
            WWWDIR => Absolute path to the SquidGuard Manager base installation
                      directory
            CGIDIR => Relative path from WWWDIR to install CGI script.
                      Default equals WWWDIR
            HTMLDIR => Relative path from WWWDIR to install css, js and
                       image files. Default equals WWWDIR
            LANGDIR => Translation language to use (en_US or fr_FR).
                       Default: en_US
            BASEURL => Path part of the base url to access SquidGuard
                       Manager Web GUI. Default: /squidguardmgr
            SQUIDUSR => User running Squid
            SQUIDGRP => Group running Squid
            SQUIDCLAMAV => Absolute path to the squidclamav binary (v5.x only).
                           Set to 'off' if you won't manage neither v5 nor V6.x
            CICAP_SOCKET => c-icap server control socket (v6.x only)
            SQUIDCLAMAVCONF => Absolute path to squidclamav.conf file

    For packaging there's two more configuration variables:

            DESTDIR => Directory prefix where install will be done
            QUIET => Specify with value '1' to suppress post install messages

    Example:

            perl Makefile.PL \
            CONFFILE=/usr/local/etc/squidguard.conf \
            DBHOME=/usr/local/squidguard/db \
            LOGDIR=/usr/local/logs/squidguard \
            WWWDIR=/usr/local/apache \
            CGIDIR=cgi-bin/admin HTMLDIR=htdocs/admin \
            LANGDIR=fr_FR BASEURL=/admin \
            SQUIDUSR=squid SQUIDGRP=squid \
            SQUIDCLAMAV=/usr/local/bin/squidclamav \
            SQUIDCLAMAVCONF=/etc/squidclamav.conf \
            QUIET=1

            make
            su
            make install

    This will install SquidGuard Manager CGI squidguardmgr.cgi,
    squidguardmgr.conf, squid_wrapper and all language files under
    /usr/local/apache/cgi-bin/admin/. All SquidGuard Manager css, js and
    image files will be copied under the directory
    /usr/local/apache/htdocs/admin/. Squidguard Manager will expect to find
    the SquidGuard configuration file at /usr/local/etc/squidguard.conf and
    the log and SquidGuard databases respectively at
    /usr/local/logs/squidguard and /usr/local/squidguard/db. All the
    installed files will be owned by user squid and group squid. The squid
    wrapper (used to reconfigure squid after change) will be setuid root
    unconditionally.

  Manual installation
    Copy all files from the cgi-bin repository into the CGI script directory
    of your choice. Check that squidguardmgr.cgi can be executable by your
    Apache server and chown the entire directory to the user and group
    running httpd.

    Copy all files from the htdocs repository into your DocumentRoot
    directory or any other subdirectory and check that they can be read by
    your httpd server.

    Edit the squidguardmgr.conf file in your cgi directory and check that
    all paths conforms to your installation.

    If you want to be able to reload Squid redirector from the GUI interface
    after your SquidGuard and/or SquidClamav modification, you must compile
    and install the squid_wrapper with suid root into the squidguardmgr cgi
    directory.

    Before compilation you may need to edit squid_wrapper.c and change the
    SQUID_BINARY definition to point correctly to your squid binary.

    Then as root:

            cd squid_wrapper/
            cc -o /path/squidguardmgr/cgi/squid_wrapper squid_wrapper.c
            chown root.root /path/squidguardmgr/cgi/squid_wrapper
            chmod 755 /path/squidguardmgr/cgi/squid_wrapper
            chmod u+s /path/squidguardmgr/cgi/squid_wrapper

CONFIGURATION
    SquidGuard Manager needs some configuration. They are stored in the
    squidguardmgr.conf file. During the automatic installation this file
    will be generated automatically.

    SQUIDGUARD /usr/local/squidGuard/bin/squidGuard
        This directive allows you to configure the path to the SquidGuard
        program. If you set it to off or no, the SquidGuard Manager
        interface will be disabled and it will only show the SquidClamav
        Manager interface.

    CONF_FILE /usr/local/squidGuard/squidguard.conf
        Configure the path to the SquidGuard configuration file.

    SQUIDCLAMAV /usr/bin/squidclamav | c-icap
        Path to the SquidClamav program when using SquidClamav v5.x branch
        and must be set to 'c-icap' when you're using v6.x branch.

    C_ICAP_SOCKET /var/run/c-icap/c-icap.ctl
        Path to the c-icap control socket. Used by SquidClamav Manager to
        reconfigure the c-icap server to apply SquidClamav configuration
        changes. Used only with SquidClamav 6.x branch.

    SC_CONF_FILE /etc/squidclamav.conf
        Path to the SquidClamav configuration file.

    LANG en_US
        Used to set the language, default is en_US. Current available LANGs
        are: en_US, fr_FR.

    SQUID_WRAPPER /var/www/squidguargmgr/squid_wrapper
        Command wrapper to reload Squid. Why you need this wrapper is
        because only root can do it. This wrapper will execute the following
        system command as root user:

                /path/to/squid/bin/squid -k reconfigure

    IMG_DIR /squidguargmgr/images
        Path part of SquidGuard Manager URL where images are found.

    CSS_FILE /squidguargmgr/squidguardmgr.css
        SquidGuard Manager URL to point to the Style Sheet(CSS) file.

    JS_FILE /squidguargmgr/squidguardmgr.js
        SquidGuard Manager URL which points to Javascript(js) file.

    LOG_LINES 1000
        Number of last lines out of log files to display.

    DNSBL your.preferred.blacklist.com,other.preferred.blacklist.com
        Comma separated list of DNS Blacklists. If you have SquidGuard 1.5
        or above put here the DNSBLs you want to use.

    TAIL, FIND, DIFF, RM, GREP
        These optional configuration directives are used to override the
        default path to the system commands /usr/bin/tail, /usr/bin/find,
        /usr/bin/diff, /usr/bin/rm and /usr/bin/grep.

    KEEP_DIFF
        This enables record of added/removed items from squidGuard
        blacklists in diff files which can be used to recover changes after
        a fresh download of blacklists. Default value is 1, enabled. Set it
        to 0 to disable this feature.

AUTHORS

ChangeLog
Fri Jan 23 2015 - v1.14

This release allow directory recursion in autocreate filters to support
blacklist database with subdirectories like shalla lists. It also allow
permanent auto-creation of filters so that new folders in blocklist directory
are auto-detected and can be automatically used to create a new filter.

Here is the complete list of change and bug fixes:

	- Fix parameter with doublequote in Sources Edition. Thanks to
	  marcohald for the report.
	- Fix case where lines in the rules/classes was doubled when
	  using Autocreate filters. Thanks to e-gaulue for the report.
	- Allow permanent autocreation of filters so that new folders
	  in squidGuard blocklist (/var/lib/squidguard/db) are auto-
	  detected and automatically create a new filter.
	- Keep empty rules in squidguard configuration file. Thanks to
	  e-gaulue for the report.
	- Allow directory recursion in autocreate filters to support sub-
	  directories in black lists. Thanks to e-gaulue for the feature
	  request.
	- Add RPM spec file. Thanks to Tatsuya-Nonogaki for the patch.

UPGRADE: Replacing squidguardmgr.cgi should be enough.

Sun Dec 22 2013 - v1.13

This release adds better support to SquidGuard management, adds some new interfaces and capabilities in blocklists or sources management. Some more bugs fix.

	- Fix some translation issues. Thanks to Travis L. Bean for the patch.
	- Added Squid Restart confirmation. Popup a JavaScript msg box before
	  actually restart Squid. Patch from Tatsuya Nonogaki.
	- Added "else" block deletion icon to ACLs Configuration page for
	  uniform operability with those of Filters, Sources and so on. Patch
	  from Tatsuya Nonogaki.
	- Allow sources to be ordered by the user using the up and down arrows
	  links. Patch from Gilles Darold.
	- Log content escaping and CSS change to not expands table beyond the
	  frame. Patch from Tatsuya Nonogaki.
	- Allow selection of "All the time" to remove time constraint in
	  edit_sources and edit_rewrites. Patch from Tatsuya Nonogaki.
	- Lot of cosmetic adjustments and code cleaning. Patches from Tatsuya
	  Nonogaki.
	- Big changes in Lists Management page. Patches from Tatsuya Nonogaki.
	- Made list names in Filter Configuration page hyperlinks to their list
	  edition page. Patch from Tatsuya Nonogaki.
	- Had capability to clear blacklists' alias and descriptions. Patch from	  Tatsuya Nonogaki.
	- Fix "Remove all" icon button didn't remove empty src block from
	  squidGuard.conf. Now removes the src block if the element in question
	  was the last one of the source group. The last element of a source
	  group still in use denies deletion. Patches from Tatsuya Nonogaki.
	- Add capability to newly create source list file. Patch from Tatsuya
	  Nonogaki.
	- Fix rare but potential db file damage when tmp file cannot be put out.
	  Added description to show_logs help content. Patch from Tatsuya
	  Nonogaki.
	- Logrotate saves old log file with gunzip per EPEL squidGuard pkg
	  default. It's better to prevent binary files from being clicked upon.
	  Patch from Tatsuya Nonogaki.
	- Update documentation and add Tatsuya Nonogaki to the authors list.
	- Fix wrapper chmod arguments in Makefile.PL. Patch from Tatsuya
	  Nonogaki.
	- Fix lot of typo in English lang file. Patch from Tatsuya Nonogaki.
	- File editor window width calibration. Patch from Tatsuya Nonogaki.
	- Use hash reference instead of global hash variables. Patch from
	  Tatsuya Nonogaki.
	- Redirect Picker table width fix. Completely disabled cell wordwrap to
	  fit the table within the popup window. Patch from Tatsuya Nonogaki.
	- Added checkbox to explicitly delete else clause from the ACL. Patch
	  from Tatsuya Nonogaki.
	- Fix squidguardmgr.css 'header' class missing semicolon. Patch from
	  Tatsuya Nonogaki.
	- Fix typo where chown command was replaced by chmod in documentation.
	  Thanks to Raul Sobon for the report.
	- Added functionality to input Redirect URLs from a popup window which
	  shows URLs used somewhere in the squidGuard.conf. Changed cgi and js.
	  Patch from Tatsuya Nonogaki.
	- Remove special windows character ^M from the code to avoid diff change
	  on each pull requests. Patch from Gilles Darold.
	- Fix impossible set of ALL for else in ACLs. Thanks to Thomas Benoit
	  for the patch.
	- Configuration is now displayed asis, without any reordering by
	  squidguardmgr.cgi. Patch from Tatsuya Nonogaki.
	- Add escapeHTML to redirect URLs. Patch from Tatsuya Nonogaki.
	- Fixed bug that inserts unnecessary empty 'else' clause into
	  squidGuard.conf. Patch from Tatsuya Nonogaki.
	- Add init parameter file for squidguardmgr httpd daemon. Based on
	  /etc/sysconfig/httpd distributed with CentOS 6.4. Patch from Tatsuya
	  Nonogaki.
	- Change CGI directory setting include-file for squidguardmgr httpd
	  daemon. Patch from Tatsuya Nonogaki.
	- Add EL6 sample httpd configuration for squidguardmgr. Patch from
	  Tatsuya Nonogaki.
	- Blacklist text and db files should be kept non-world-readable.
	  Insertion of umask(). Patch from Tatsuya Nonogaki.
	- Fix source list file edition that was not working properly via IE or
	  Firefox on Windows machines. Patch from Tatsuya Nonogaki.
	- Not all ACL objects require "redirect" unless it also has "log" set.
	  Patch from Tatsuya Nonogaki.
	- Fix "log" setting for each ACL block in squidGuard.conf that have to
	  be parsed. Patch from Tatsuya Nonogaki.
	- Replace direct call to squidguard binary to reload db files by the
	  call to the rebuild_database() function. Patch from Gilles Darold.

UPGRADE: All files have changes, you must fully override your previous
installation by using: perl Makefile.PL && make && make install.

Sun Jan 27 2013 - v1.12

	- Add more debug information about squidGuard.conf and squidclamav.conf
	  configuration file when squidguardmgr can not write to it.
	- Fix delete error of items in squidclamav virus scanning, whitelists
	  and trusted users definitions. Thanks to John Xue for the report.

Makefile.PL
use ExtUtils::MakeMaker qw(prompt WriteMakefile);

my @ALLOWED_ARGS = ('SQUIDGUARD','CONFFILE','DBHOME','CGIDIR','HTMLDIR','DESTDIR','QUIET','LANGDIR','BASEURL','WWWDIR','LOGDIR','SQUIDUSR','SQUIDGRP','SQUIDCLAMAV','CICAP_SOCKET','SQUIDCLAMAVCONF');

# Parse command line arguments and store them as environment variables
while ($_ = shift) {
	my ($k,$v) = split(/=/, $_, 2);
	if (grep(/^$k$/, @ALLOWED_ARGS)) {
		$ENV{$k} = $v;
	}
}
$ENV{DESTDIR} =~ s/\/$//;

# Default install path for packaging support.
my $DESTDIR = $ENV{DESTDIR} || '';
$ENV{INSTALLDIRS} ||= 'vendor';
# Default install path
my $SQUIDGUARD = $ENV{SQUIDGUARD} || '';
my $CONFFILE = $ENV{CONFFILE} || ''; # SquidGuard configuration file
my $DBHOME = $ENV{DBHOME} || ''; # SquidGuard DB directory
my $LOGDIR = $ENV{LOGDIR} || ''; # SquidGuard logs directory
my $LANGDIR = $ENV{LANGDIR} || 'en_US'; # Default SquidGuard Manager language
# SquidGuard Manager default install base directory
my $WWWDIR = $ENV{WWWDIR} || '/var/www/squidguardmgr';
$WWWDIR .= '/' if ($WWWDIR && ($WWWDIR !~ /\/$/));
# Where to install CGI: $WWWDIR/$CGIDIR. Default: $WWWDIR
my $CGIDIR = $ENV{CGIDIR} || '';
$CGIDIR .= '/' if ($CGIDIR && ($CGIDIR !~ /\/$/));
# Where to install css,js and images files: $WWWDIR/$HTMLDIR. Default: $WWWDIR
my $HTMLDIR = $ENV{HTMLDIR} || '';
$HTMLDIR .= '/' if ($HTMLDIR && ($HTMLDIR !~ /\/$/));
# Base url where css,js and images files will be found. Default: /squidguardmgr
my $BASEURL = $ENV{BASEURL} || '/squidguardmgr';
# Squid running user and group
my $SQUIDUSR = $ENV{SQUIDUSR} || '';
my $SQUIDGRP = $ENV{SQUIDGRP} || '';
my $CICAP_SOCKET = $ENV{CICAP_SOCKET} || '';
my $SQUIDCLAMAVCONF = $ENV{SQUIDCLAMAVCONF} || '';
my $SQUIDCLAMAV = $ENV{SQUIDCLAMAV} || '';
if ($CICAP_SOCKET) {
	$SQUIDCLAMAV = 'c-icap';
}

# Try to find all binary used by SquidGuard Manager
$WWWDIR .= '/' if ($WWWDIR && ($WWWDIR !~ /\/$/));
my $tail = `which tail 2>/dev/null`;
chomp($tail);
if (!$tail) {
	die "NOTICE: This tool requires system command: tail, can't find it in your path\n";
}
my $grep = `which grep 2>/dev/null`;
chomp($grep);
if (!$grep) {
	die "NOTICE: This tool requires the system command: grep, can't find it in your path\n";
}
my $find = `which find 2>/dev/null`;
chomp($find);
if (!$find) {
	die "NOTICE: This tool requires the system command: find, can't find it in your path\n";
}
my $diff = `which diff 2>/dev/null`;
chomp($diff);
if (!$diff) {
	die "NOTICE: This tool requires the system command: diff, can't find it in your path\n";
}
my $rm = `which rm 2>/dev/null`;
chomp($rm);
if (!$rm) {
	die "NOTICE: This tool requires the system command: rm, can't find it in your path\n";
}

# Get the path to squidGuard binary or disable it
my $squidGuard = $SQUIDGUARD || '';
my $squidGuardConf = $CONFFILE || '';
if (lc($squidGuard) eq 'off') {
	$squidGuard = 'off';
	$squidGuardConf = '';
} else {
	if (!$squidGuard) {
		$squidGuard = `which squidGuard 2>/dev/null`;
		chomp($squidGuard);
	}
	if (!-e $squidGuard) {
		$squidGuard = prompt('Path to squidGuard program?','off');
	}
	if (!-e $squidGuard && (lc($squidGuard) ne 'off')) {
		die "NOTICE: This tool requires the system command: squidGuard, can't find it in your path\n";
	}
	if (!$squidGuardConf) {
		foreach ('/usr/local/squidGuard/squidguard.conf','/usr/local/squidGuard/etc/squidguard.conf','/usr/local/etc/squidguard.conf','/etc/squidguard.conf','/etc/squid/squidguard.conf','/etc/squid/squidGuard.conf') {
			$squidGuardConf = $_, last if (-e $_);
		}
		if (lc($squidGuard) eq 'off') {
			$squidGuard = 'off';
		} else {
			if (!-e $squidGuardConf) {
				$squidGuardConf = prompt('Path to squidGuard configuration file?', '');
			}
			if (!-e $squidGuardConf) {
				print "NOTICE: can't find squidguard.conf in your system.\n";
			}
		}
	}
	if (-e $squidGuardConf) {
		if (!$DBHOME) {
			my $dbhome = `grep "dbhome" $squidGuardConf`;
			chomp($dbhome);
			if ($dbhome =~ s/^[^\s\t]+[\s\t]+//) {
				$DBHOME = $dbhome;
			}
		}
		if (!$LOGDIR) {
			my $logdir = `grep "logdir" $squidGuardConf`;
			chomp($logdir);
			if ($logdir =~ s/^[^\s\t]+[\s\t]+//) {
				$LOGDIR = $logdir;
			}
		}
	}
}

my $squidclamav = $SQUIDCLAMAV || '';
my $cicapSocket = $CICAP_SOCKET || '';
my $squidclamavConf = $SQUIDCLAMAVCONF || '';
if (lc($squidclamav) eq 'off') {
	$squidclamav = 'off';
	$cicapSocket = '';
	$squidclamavConf = '';
} elsif (!$cicapSocket && !$squidclamav) {
	$squidclamav = prompt('Do you use c-icap SquidClamav v6.x services?', 'no');
	if ($squidclamav && $squidclamav !~ /^no/i) {
	# Possibility is invoker want to use SquidClamav 6.x
		$squidclamav = 'c-icap';
		# Search for the c-icap command socket
		my $cicapConf = '';
		foreach ('/usr/local/c-icap/etc/c-icap.conf','/usr/local/etc/c-icap.conf','/etc/c-icap/c-icap.conf','/etc/c-icap.conf') {
			$cicapConf = $_, last if (-e $_);
		}
		if (!$cicapConf) {
			$cicapConf = prompt('Path to c-icap configuration file?', '/usr/local/etc/c-icap.conf');
		}
		if (-e $cicapConf) {
			$cicapSocket = `grep "^CommandsSocket" $cicapConf | awk '{print \$2}'`;
			chomp($cicapSocket);
		}
	} else {
	# Possibility is invoker wants to use SquidClamav 5.x or none
		$squidclamav = `which squidclamav 2>/dev/null`;
		chomp($squidclamav);
		if (!-e $squidclamav) {
			$squidclamav = prompt('Path to squidclamav program?', 'off');
			if (!-e $squidclamav && (lc($squidclamav) ne 'off')) {
				die "NOTICE: This tool requires the system command: squidclamav, can't find it in your path\n";
			}
			if (lc($squidclamav) eq 'off') {
				$squidclamav = 'off';
			}
		}
	}
} elsif ($cicapSocket) {
	$squidclamav = 'c-icap';
} else {
	# Ensure user specified squidclamav existence
	if (!-e $squidclamav) {
		$squidclamav = prompt('Path to squidclamav program?', 'off');
		if (!-e $squidclamav && (lc($squidclamav) ne 'off')) {
			die "NOTICE: This tool requires the system command: squidclamav, can't find it in your path\n";
		}
		if (lc($squidclamav) eq 'off') {
			$squidclamav = 'off';
		}
	}
}

if ( ($squidclamav eq 'off') && ($squidGuard eq 'off')) {
	die "FATAL: you must either have squidGuard or SquidClamav installed!\n";
}

if ( ($squidclamav ne 'off') && !$squidclamavConf) {
	foreach ('/usr/local/etc/squidclamav.conf','/etc/squidclamav.conf','/etc/squidclamav/squidclamav.conf') {
		$squidclamavConf = $_, last if (-e $_);
	}
	if (!-e $squidclamavConf) {
		$squidclamavConf = prompt('Path to SquidClamav configuration file?', '');
		if (!-e $squidclamavConf) {
			die "NOTICE: can't find squidclamav.conf in your system\n";
		}
	}
}

my $squid = `which squid 2>/dev/null`;
chomp($squid);
$squid ||= `which squid3 2>/dev/null`;
chomp($squid);
$squid ||= '/usr/local/squid/sbin/squid';
if (!-e $squid) {
	$squid = prompt('Path to squid program?', '');
	if (!-e $squid && (lc($squid) ne 'off')) {
		die "NOTICE: This tool requires the system command: squid, can't find it in your path\n";
	}
}
$squid =~ s/\//\\\//g;
# Change default path to squid binary
`perl -p -i -e 's/#define SQUID_BINARY.*/#define SQUID_BINARY "$squid"/' squid_wrapper/squid_wrapper.c`;


my $squid_user = $SQUIDUSR || `ps auxw | grep "(squid)" | grep -v grep | awk '{print \$1}'`;
chomp($squid_user);
my $squid_group = $SQUIDGRP || $squid_user;

if (!$squid_user) {
	$squid_user = prompt('What is the user (uid) running Squid or SquidGuard?', 'squid');
	my $ret = `id $squid_user 2>/dev/null`;
	chomp($ret);
	if (!$ret) {
		die "NOTICE: SquidGuard Manager must be run as SquidGuard user, uid=$squid_user not found!\n";
	} else {
		if ($ret =~ /uid=(\d+)\(.* gid=(\d+)\(/) {
			$squid_user = $1;
			$squid_group = $2;
		}
	}
	if (!$squid_user) {
		die "NOTICE: SquidGuard Manager must be run as SquidGuard user, not found!\n";
	}
}
if (!$squid_group) {
	my $ret = `id $squid_user 2>/dev/null`;
	chomp($ret);
	if (!$ret) {
		die "NOTICE: SquidGuard Manager must be run as SquidGuard group, not found!\n";
	} else {
		if ($ret =~ /uid=(\d+)\(.* gid=(\d+)\(/) {
			$squid_group = $2;
		}
	}
	if (!$squid_group) {
		die "NOTICE: SquidGuard Manager must be run as SquidGuard group, not found!\n";
	}
}

# Setup ok. generating squidguardmgr.conf config file
unless(open(OUTCFG, ">cgi-bin/squidguardmgr.conf")) {
	print "\nError: can't write config file cgi-bin/squidguardmgr.conf, $!\n";
	exit 0;
}

	print OUTCFG qq{
#
# SquidGuard Manager configuration file
#

# Path to the grep system command
GREP	$grep

# Path to the find system command
FIND	$find

# Path to the tail system command
TAIL	$tail

# Path to the diff system command
DIFF	$diff

# Path to the rm system command
RM	$rm

# Path to the SquidGuard program. You can set it to off to disable the
# SquidGuard manager interface and get only the SquidClamav manager.
SQUIDGUARD	$squidGuard

# Path to the SquidGuard configuration file
CONF_FILE	$squidGuardConf

# Path to the SquidClamav program. You can set it to off to disable the
# SquidClamav manager interface and get only the SquidGuard manager.
# If you use SquidClamav v6.x with the c-icap server, set this directive to
# the value: c-icap
SQUIDCLAMAV	$squidclamav

# Path to the c-icap control socket. Used by SquidClamav Manager to reload
# c-icap server to apply SquidClamav configuration changes. Used only with
# SquidClamav v6.x branch.
C_ICAP_SOCKET	$cicapSocket

# Path to the SquidClamav configuration file
SC_CONF_FILE	$squidclamavConf

# Used to set the language, default is en_US.
# Current translation are: en_US, fr_FR.
LANG		$LANGDIR

# Command to reload Squid. You may use the wrapper as squid
# can only be reload as root. This wrapper will run the command
# /usr/local/squid/sbin/squid -k reconfigure
SQUID_WRAPPER	${WWWDIR}${CGIDIR}squid_wrapper

# SquidGuardMgr URL base where images are stored
IMG_DIR		$BASEURL/images

# SquidGuardMgr Style Sheet CSS URL
CSS_FILE	$BASEURL/squidguardmgr.css

# SquidGuardMgr Javascript URL
JS_FILE		$BASEURL/squidguardmgr.js

# Comma separated list of DNS Blacklists
#DNSBL	your.preferred.blacklist.com,other.preferred.blacklist.com

# Number of last lines displayed from log files
TAIL_LINES	1000

# Keep added/removed items from squidGuard blocklists as diff file
# to be able to recover change after a fresh download of blocklists
KEEP_DIFF       1

};
close(OUTCFG);

unless(open(INST, ">install_all.sh")) {
	print "\nError: can't write post install file install_all.sh, $!\n";
	exit 0;
}
print INST qq{#!/bin/sh
if [ ! -d "$DESTDIR$WWWDIR" ]; then
	mkdir -p $DESTDIR$WWWDIR
fi

if [ -e $DESTDIR$WWWDIR${CGIDIR}squidguardmgr.conf ]; then
	echo "File squidguardmgr.conf already exists. Renaming it to squidguardmgr.conf.old"
	mv $DESTDIR$WWWDIR${CGIDIR}squidguardmgr.conf $DESTDIR$WWWDIR${CGIDIR}squidguardmgr.conf.old
fi
};
if ($DBHOME) {
	print INST qq{
perl -p -i -e 's#my \\\$DBHOME.*=.*#my \\\$DBHOME = "$DBHOME";#' cgi-bin/squidguardmgr.cgi
};
}

if ($LOGDIR) {
	print INST qq{
perl -p -i -e 's#my \\\$LOGDIR.*=.*#my \\\$LOGDIR = "$LOGDIR";#' cgi-bin/squidguardmgr.cgi
};
}

print INST qq{
install -m 755 --owner=$squid_user --group=$squid_group cgi-bin/squidguardmgr.cgi $DESTDIR$WWWDIR$CGIDIR
install -m 644 --owner=$squid_user --group=$squid_group cgi-bin/squidguardmgr.conf $DESTDIR$WWWDIR$CGIDIR
cp -rf cgi-bin/lang/ $DESTDIR$WWWDIR$CGIDIR
chown -R $squid_user:$squid_group $DESTDIR$WWWDIR$CGIDIR
install -m 644 --owner=$squid_user --group=$squid_group htdocs/squidguardmgr.css $DESTDIR$WWWDIR$HTMLDIR
install -m 644 --owner=$squid_user --group=$squid_group htdocs/squidguardmgr.js $DESTDIR$WWWDIR$HTMLDIR
cp -rf htdocs/images/ $DESTDIR$WWWDIR$HTMLDIR
chown -R $squid_user:$squid_group $DESTDIR$WWWDIR$HTMLDIR
install -m 755 --owner=0 --group=0 squid_wrapper/squid_wrapper $DESTDIR$WWWDIR$CGIDIR
chmod u+s $DESTDIR$WWWDIR$CGIDIR/squid_wrapper
};
if (!$ENV{QUIET}) {
	print INST qq{
echo "
-----------------------------------------------------------------------------

1. Modify Apache httpd.conf to change the running uid and gid to squid user.
2. Modify your httpd.conf to allow access to CGI scripts like follow:

	Alias /squidguardmgr $WWWDIR$HTMLDIR
	<Directory $WWWDIR$HTMLDIR>
	    Options +ExecCGI
	    AddHandler cgi-script .cgi
	    DirectoryIndex squidguardmgr.cgi
	    Order deny,allow
	    Deny from all
	    Allow from 192.168.1.24
	</Directory>

3. If necessary, set additional grants to Squidguard Manager in httpd.conf.
   Restart and ensure that httpd is running.
4. Browse to http://proxy.host.dom/squidguardmgr/ to ensure that things are
   working properly.
For more information, see http://squidguardmgr.darold.net/.
-----------------------------------------------------------------------------
" | more
};
}
close(INST);

`chmod 755 install_all.sh`;

WriteMakefile(
    'NAME'         => 'squidguardmgr',
    'VERSION_FROM' => 'cgi-bin/squidguardmgr.cgi',
    'dist'         => {
			'COMPRESS'=>'gzip -9f', 'SUFFIX' => 'gz',
			'ZIP'=>'/usr/bin/zip','ZIPFLAGS'=>'-rl'
		      },
    'AUTHOR'       => 'Gilles Darold (gilles@darold.net)',
    'ABSTRACT'     => 'SquidGuard/SquidClamav administration GUI',
    'DESTDIR'      => $ENV{DESTDIR},
    'INSTALLDIRS'  => $ENV{INSTALLDIRS},
    'clean'        => {FILES => "install_all.sh cgi-bin/squidguardmgr.conf squid_wrapper/squid_wrapper"}
);

sub MY::install {
	my $self = shift;

	my $string = $self->MM::install;
	#$string =~ s/(pure_install\s+)(.*)/$1 install_all $2/;
	$string =~ s/pure_install\s+.*/install_all/;

	return $string;
}

sub MY::postamble {
	my $postamble = <<'END';
install_all: install_all.sh squid_wrapper/squid_wrapper.c
	cc -o squid_wrapper/squid_wrapper squid_wrapper/squid_wrapper.c
	sh install_all.sh
END
	return $postamble;
}

if (!$ENV{QUIET}) {
	print "Done...\n\n";
	print "Now type 'make && make install'\n\n";
}

